exporter/metric/prometheus/prometheus.go

```
@@ -0,0 +1,262 @@
+	DefaultSummaryObjectives map[float64]float64
+
+	// OnRegisterError defines a method to call to when registering
+	// a metric with the registerer fails. Use nil to specify
+	// to panic by default when registering fails.
+	OnRegisterError func(err error)
+}
+
+// NewExporter returns a new prometheus exporter for prometheus metrics.
+func NewExporter(opts Options) (*Exporter, error) {
+	if opts.Registerer == nil {
+		opts.Registerer = prometheus.DefaultRegisterer
+	} else {
+		// A specific registerer was set, check if it's a registry and if
+		// no gatherer was set, then use that as the gatherer
+		if reg, ok := opts.Registerer.(*prometheus.Registry); ok && opts.Gatherer == nil {
+			opts.Gatherer = reg
+		}
+	}
+	if opts.Gatherer == nil {
+		opts.Gatherer = prometheus.DefaultGatherer
+	}
+	if opts.OnRegisterError == nil {
+		opts.OnRegisterError = func(err error) {
+			panic(err)
+		}
+	}
+
+	// TODO: should we make a "PullController" ?
+	go func() {
+		_ = http.ListenAndServe(":22022", promhttp.HandlerFor(opts.Gatherer, promhttp.HandlerOpts{}))
+	}()
+
+	return &Exporter{
+		registerer:      opts.Registerer,
+		gatherer:        opts.Gatherer,
+		onRegisterError: opts.OnRegisterError,
+
+		counters: make(map[metricID]prometheus.Counter),
+		gauges:   make(map[metricID]prometheus.Gauge),
+
+		counterVecs: make(map[string]*prometheus.CounterVec),
+		gaugeVecs:   make(map[string]*prometheus.GaugeVec),
+	}, nil
+}
+
+// Export exports the provide metric record to prometheus.
+func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+	var aggError error
+	checkpointSet.ForEach(func(record export.Record) {
+		agg := record.Aggregator()
+		if sum, ok := agg.(aggregator.Sum); ok {
+			value, err := sum.Sum()
+			if err != nil {
+				aggError = err
+				// TODO: halp, where are we doing here?
+				return
+			}
+
+			c, err := e.getCounter(record)
+			if err != nil {
+				// TODO: log a warning here?
```

This seems pretty serious. I'd like to update the metrics API with a specification that can avoid errors of this nature. Perhaps we should specify a size-limit for metric names, valid character set for keys, and so on. I'd be fine with TODOs, but if the Exporter returns an error the SDK will at least call the error handler, which is better than swallowing the error here.

---

+1, agreed.. With the SDK allowing returning errors now, that's how we should deal with it.